File: analyzer/src/ast/expressions/identifier_expression.rs
use crate::{
    ast::identifier::Identifier, error::Result, lexer::token_type::TokenType, parser::Parser,
};

use super::{Expression, Precedence};

#[derive(Debug)]
pub struct IdentifierExpression {
    pub id: Identifier,
}

impl IdentifierExpression {
    pub fn new(id: Identifier) -> Self {
        Self { id }
    }

    pub fn parse(parser: &mut Parser, precedence: Precedence) -> Result<Expression> {
        let token = parser.current();
        if token.token_type() != &TokenType::Identifier {
            return Expression::parse_expr(parser, precedence.next());
        }
        let id = Identifier::from(token.text());
        return Ok(Expression::Identifier(IdentifierExpression::new(id)));
    }
}

File: src/Parser/Expressions/IdentifierExpression.cpp
#include "Parser/Parser.hpp"

#include "Ast/Expressions.hpp"

namespace Marble
{

    IdentifierExpression::IdentifierExpression(const Identifier &identifier, const Span &span)
        : Expression{span, ExpressionType::Identifier}, m_Identifier{identifier}
    {
    }

    IdentifierExpression::IdentifierExpression(Identifier &&identifier, Span &&span)
        : Expression{std::move(span), ExpressionType::Identifier}, m_Identifier{std::move(identifier)}
    {
    }

    IdentifierExpression::IdentifierExpression(const IdentifierExpression &obj)
        : Expression{obj.m_Span, ExpressionType::Identifier}, m_Identifier{obj.m_Identifier}
    {
        if (obj.m_ValueType)
        {
            m_ValueType = MakeRef<TypeSpecifier>(*obj.m_ValueType.get());
        }
    }

    Box<Expression> IdentifierExpression::Parse(Parser &parser, Precedence precedence)
    {
        const Token &token = parser.Current();
        if (token.TokenType() != TokenType::Identifier)
        {
            return Expression::Parse(parser, Expression::NextPrecedence(precedence));
        }
        return std::make_unique<IdentifierExpression>(Identifier{token.Text(), token.Span()}, token.Span());
    }
} // namespace Marble
